app/uk/gov/hmrc/gform/gform/processor/FormProcessor.scala

```diff
@@ -78,7 +80,24 @@ class FormProcessor(
     lang: LangADT,
     sse: SmartStringEvaluator
   ): Future[Result] = {
-
+    def computePageLink(
+      formPageId: PageId,
+      pageIdSectionNumberMap: Map[ModelPageId, SectionNumber]
+    ) = {
+      val forModelPageId = formPageId.modelPageId
+      pageIdSectionNumberMap.get(forModelPageId) match {
+        case Some(sectionNumber) => Some(sectionNumber)
+        case None =>
```

Comment:
What is happening in the body of this `None` case? I can't understand that from reading the code.

Reply:
> What is happening in the body of this `None` case? I can't understand that from reading the code.

to be honest I've stolen it from the method that compute the link LinkCtx expression
https://github.com/hmrc/gform-frontend/blob/0e3ada588371e1a4d0fa82e568ff570984b3080f/app/uk/gov/hmrc/gform/eval/EvaluationResults.scala#L499

from what I understand if it can't find a pageId in the map it tries to find the first page with the same basePageId

Reply:
So this is AC4 I assume, see https://github.com/hmrc/gform-frontend/pull/1653#discussion_r1020355493 for better alternative

Reply:
> So this is AC4 I assume, see [#1653 (comment)](https://github.com/hmrc/gform-frontend/pull/1653#discussion_r1020355493) for better alternative

Not it is not AC4. AC4 works almost out of the box when I change that:
```

Reply:
What AC is it then?

Reply:
This is part of AC2. The case with "pageIdToDisplayAfterRemove": "addShipping". 
`addSheeing` is a `pageId`. I tested a simple case when

```
  {
      "_id": "form1972-with-default",
      "formName": "form 1972 with default",
      "description": "Add to list",
      "authConfig": {
          "authModule": "anonymous"
      },
      "emailTemplateId": "eeitt_submission_confirmation",
      "sections": [
             {
               "title": "Which country will you start your journey from?",
               "presentationHint": "invisiblePageTitle",
               "id": "page1",
                "fields": [
                  {
                    "id": "fid1",
                    "type": "text",
                    "format": "text",
                    "label": "First name"
                  }
                ]
             },
            {
              "id": "listOfDirectors",
              "type": "addToList",
              "title": "ATL title",
              "summaryName": "Directors",
              "shortName": "Director $n",
              "AAdescription": "**$n.**",
              "description": "",
              "summaryDescription": " asdfa sdfa sdf asdf as fda fds${directorName}",
              "pageIdToDisplayAfterRemove": "page1",
              "addAnotherQuestion": {
                "id": "directorList",
                "type": "choice",
                "label": "Add another director?",
                "format": "yesno",
                "errorMessage": "Select whether you want to add another director"
              },
              "defaultPage": {
                "title": "Add details default page",
                "fields": [
                  {
                    "id": "dp1",
                    "label": "",
                    "type": "info",
                    "infoType": "noformat",
                    "infoText": "You must tell us about all directors default page"
                  }
                ]
              },
              "cyaPage": {
                "updateTitle": "Change answers for vehicle $n",
                "noPIIUpdateTitle": "Change answers for vehicle $n",
                "title": "Check your answers for vehicle $n",
                "noPIITitle": "Check your answers for vehicle $n"
              },
              "pages": [
                {
                  "progressIndicator": "Director $n",
                  "title": "ATL Page 1 title ",
                  "presentationHint": "invisiblePageTitle",
                  "fields": [
                    {
                      "id": "directorsInfo",
                      "type": "info",
                      "label": "director info",
                      "infoText": "Add the name of the primary director, other directors can be added later.",
                      "infoType": "noformat"
                    },
                    {
                      "id": "directorName",
                      "type": "text",
                      "format": "shortText",
                      "label": "director name",
                      "labelSize": "s",
                      "shortName": "Director name"
                    },
                    {
                      "id": "directorSurname",
                      "type": "text",
                      "format": "shortText",
                      "label": "director surname",
                      "labelSize": "s",
                      "shortName": "Director surname"
                    }
                  ]
                },
                {
                  "progressIndicator": "Page $n",
                  "title": "ATL Page 2 title",
                  "presentationHint": "invisiblePageTitle",
                  "fields": [
                    {
                      "id": "foobar2",
                      "type": "text",
                      "format": "shortText",
                      "label": "director surname2",
                      "labelSize": "s",
                      "shortName": "Director surname Page 2"
                    }
                  ]
                }
              ]
            }
      ],
      "declarationSection": {
          "shortName": "Declaration",
          "title": "Declaration",
          "fields": []
      },
      "acknowledgementSection": {
          "shortName": "Acknowledgement Page",
          "title": "Acknowledgement Page",
          "fields": []
      },
      "destinations": [
          {
              "id": "transitionToSubmitted",
              "type": "stateTransition",
              "requiredState": "Submitted"
          }
      ]
  }
```

Reply:
I will try to come up with a test hitting a second `if` branch on Monday.

Reply:
here is the form with 2 ATLs. The second ATL refers to the first one. And when we delete the last iteration of the second ATL it jumps to the first one. In that case the `if` branch with `              modelPageId.baseId === forModelPageId.baseId
` is triggered
https://gist.github.com/lxol/f868aab37f97e8ccb332c54a9af53b19

Reply:
Ok, can you please add comment to that code handling reference to second ATL.

And also please do something with unsafe calls: https://github.com/hmrc/gform-frontend/pull/1653/files#r1021356900

Reply:
done